File: lib/api/models/task.dart
// lib/models/task.dart

class Task {
  final int? taskId;
  final String title;
  final String? description;
  final int? createdByUserId;
  final DateTime? createdAt;
  final DateTime? lastUpdated;
  final bool isCompleted;

  // Related entities
  final Map<String, dynamic>? user;
  final List<Map<String, dynamic>>? taskCategories;
  final List<Map<String, dynamic>>? categories;

  Task({
    this.taskId,
    required this.title,
    this.description,
    this.createdByUserId,
    this.createdAt,
    this.lastUpdated,
    this.isCompleted = false,
    this.user,
    this.taskCategories,
    this.categories,
  });

  factory Task.fromJson(Map<String, dynamic> json) {
    return Task(
      taskId: json['task_id'],
      title: json['title'],
      description: json['description'],
      createdByUserId: json['created_by_user_id'],
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : null,
      lastUpdated: json['last_updated'] != null
          ? DateTime.parse(json['last_updated'])
          : null,
      isCompleted: json['is_completed'] ?? false,
      user:
          json['user'] != null ? Map<String, dynamic>.from(json['user']) : null,
      taskCategories: (json['task_categories'] as List?)
          ?.map((e) => Map<String, dynamic>.from(e))
          .toList(),
      categories: (json['categories'] as List?)
          ?.map((e) => Map<String, dynamic>.from(e))
          .toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (taskId != null) 'task_id': taskId,
      'title': title,
      if (description != null) 'description': description,
      if (createdByUserId != null) 'created_by_user_id': createdByUserId,
      if (createdAt != null) 'created_at': createdAt!.toIso8601String(),
      if (lastUpdated != null) 'last_updated': lastUpdated!.toIso8601String(),
      'is_completed': isCompleted,
      if (user != null) 'user': user,
      if (taskCategories != null) 'task_categories': taskCategories,
      if (categories != null) 'categories': categories,
    };
  }

  Task copyWith({
    int? taskId,
    String? title,
    String? description,
    int? createdByUserId,
    DateTime? createdAt,
    DateTime? lastUpdated,
    bool? isCompleted,
    Map<String, dynamic>? user,
    List<Map<String, dynamic>>? taskCategories,
    List<Map<String, dynamic>>? categories,
  }) {
    return Task(
      taskId: taskId ?? this.taskId,
      title: title ?? this.title,
      description: description ?? this.description,
      createdByUserId: createdByUserId ?? this.createdByUserId,
      createdAt: createdAt ?? this.createdAt,
      lastUpdated: lastUpdated ?? this.lastUpdated,
      isCompleted: isCompleted ?? this.isCompleted,
      user: user ?? this.user,
      taskCategories: taskCategories ?? this.taskCategories,
      categories: categories ?? this.categories,
    );
  }
}

File: final_project/models/task.py
from datetime import datetime
from sqlalchemy import Column, Integer, String, Text, Boolean, DateTime, ForeignKey
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from ..db import Base

class TaskDB(Base):
    __tablename__ = "tasks"

    task_id = Column(Integer, primary_key=True, autoincrement=True)
    title = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)

    # Replace admin reference with user reference
    created_by_user_id = Column(
        Integer,
        ForeignKey("users.user_id", ondelete="SET NULL"),
        nullable=True
    )

    created_at = Column(
        DateTime,
        server_default=func.now(),
        nullable=False
    )
    last_updated = Column(
        DateTime,
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False
    )
    is_completed = Column(Boolean, default=False, nullable=False)

    # Relationship to UserDB
    user = relationship(
        "UserDB",
        back_populates="tasks",
        foreign_keys=[created_by_user_id]
    )

    # Other relationships
    task_categories = relationship(
        "TaskCategoryDB",
        back_populates="task",
        cascade="all, delete-orphan"
    )
    
    categories = relationship(
        "CategoryDB",
        secondary="task_category",
        back_populates="tasks"
    )
    # Add additional relationships as needed
